### DISCUSSION TITLE
Tie notes - How can we implement this with smplr?

### DISCUSSION THREAD

**Discussion by naushad-jamil** [danigb/smplr#83] [Q&A] [answered]
Let's say an app wants to play a series of notes (pitch, duration) using an instrument in smplr. Plain and simple. It works.
e.g. C4,0.5 D4,1 etc.

What if we have a series of notes but there are some tie notes. e.g.
C4,0.5 tied C4,1

Here we don't want to play two different C4 the first one 0.5 and the second one 1 but we want to continue the C4,0.5 note and extend it by another 1s.

Of course we can simply pass it as C4,1.5. This is possible to do in my parser but the logic will be convoluted as I have to do look ahead and remove notes from future measures and bring them to the current measure. It will be ugly.

Another idea - is to cancel the previous note and reschedule another C4 with 1.5s duration with smplr. Is there a way to cancel a scheduled note and reschedule it?

Is there a more straight forward approach to achieve tied notes?

**Comment by danigb** [comment]
Hi @naushad-jamil 

There's no way to cancel and reschedule notes, so I'd say that look ahead and remove notes is the simplest option I can think of.

**Reply by naushad-jamil** [reply]
Hi @danigb 

Imagine you are building a player that plays measure by measure. On a 4/4 each measure plays notes of total length of 4 quarter notes. Tracking this measure by measure in the player one can implement some visual feedback. If I parse out the notes from all measures and give an accumulated sequence of notes of pitch and duration I will lose that tracking capacity (unless I go into really spaghetti code sort of stuff). So I would like to avoid that approach if possible at all and want to see if other libraries somehow provide something for tie notes.

If I could start playing a note without supplying duration and then in a timer callback stop that playing note or simply wait some more time that could also allow me to build it in a simple way. Can this be achieved?

Here is the algorithm:

// play C4 for .5s
const stopNote = piano.start({ note:"C4"}); //this starts the note, and will continue to play it until a stop command is invoked.

Wait for .5s 
Now we have the next note to play. This can be regular note with pitch and duration or a tied note which simply indicates continuation of the last pitch for x seconds.

Let's say this is tie note and duration = 1s

I will simply wait another 1s and the previous C4 will continue to play.

If this is a regular note then I will stop the last playing note
stopNote()

and play the next note

I did a quick try with this approach and it doesn't work.

The stopNote() does not seem to stop the playing note at all.

Also - 
piano.start({ note:"C4", duration:10 });

This does not play C4 for 10 seconds. It decays much sooner.

Is there a way to play the note with same velocity indefinitely until I invoke stop?

Thank you.

**Reply by drscottlobo** [reply]
In my implementation I simply loop over any tied notes and increase the first tied note’s duration by the amount of time added by any tied notes in the same voice. Then I schedule the notes. For rests I use 0 velocity notes of the correct duration. It seems to work fine. If your notes use the onStart and onEnded callbacks you can also change the visual state to show which note is playing.

**Reply by naushad-jamil** [reply]
P.S. - Adding this after I added notes below. I think your loop technique is exactly what I was suggesting in my original post and what @danigb was also suggesting right? Your parser adds a note with duration n. Then goes to the next note and if that is a tie, it continues to add the tie duration to the current note. Is that what you were suggesting? I described the problem with this approach. We will have to consider the whole song as one big sequence of notes. Instead of measure by measure. The latter is needed to keep track which measure is being played currently if we want to provide a visual feedback to the end user.


________________ 
Hi @drscottlobo I am sure you have solved the tied note problem. Can you share some code snippet please. For rest 0 velocity idea is great. I was doing it with simply adding the duration to context.currentTime.

Here is an example input array of format pitch, duration.
const input = [{"C4", 0.5}, {"D4", 1}, {"Tie", 0.5}, {"E4", 1}]

some snippet playing/scheduling these notes would be very helpful. Note {"Tie", 0.5} this need to play the previous D4 a total of 1.5s then play the new E4.

Thank you very much.

**Reply by drscottlobo** [reply]
Hi @naushad-jamil, 

I do something like this: 
```
const music = [
	{ pitch: "C4", duration: 1 },
	{ pitch: "D4", duration: 1, isTied: true },
	{ pitch: "D4", duration: 0.5 }, //// this note will skipped in loop, but it's duration will be added to tied note.
	{ pitch: "E4", duration: 1 },
];

function scheduler(music, ctx) {
	const now = ctx.currentTime;
	let timeToStart = 0;

	for (let i = 0; i < music.length; i++) {
		const note = music[i];
		if (note.isTied) {
			const nextNote = music[i + 1];
			note.duration += nextNote.duration;
			i += 1; //// move one note forward in loop to skip over the D4 of duration 0.5
		}

		player.start({
			time: now + timeToStart,
			duration: note.duration,
			note: note.pitch,
			onStart: () => {
				highlightNote()
			},
			onEnded: () => {
				removeHighlight()
			}
		});
		timeToStart += note.duration; //// start next note after current note completes
	}
}


```

**Reply by naushad-jamil** [reply]
@drscottlobo - Hello Scott. Thank you so much for taking the trouble to put together a snippet. Really appreciate it. In my original post I said
"Of course we can simply pass it as C4,1.5. This is possible to do in my parser but the logic will be convoluted as I have to do look ahead and remove notes from future measures and bring them to the current measure. It will be ugly."

This is exactly what I was referring to. That is look ahead and bring the duration to the current note in an iterative fashion.

My only problem with this approach is this -

Imagine I have UI notes editor. It is a sequence of measures/bars. I was hoping to highlight each bar (each note inside each bar would be an added bonus) as the player is playing the notes of. Imagine a 4/4 notation. bar 1's last quarter note is tied with bar 2's first quarter note. How will I highlight bar 2 when the player is playing the tied note that falls in bar 2?

That would be my challenge with this approach. 

Please let me know if you have any idea.

I thank you however, for showing me the onStart() and onEnded() callbacks. I didn't know they existed. I am sure I can make good use of them.

Again - Thanks a ton!

P.S.
I tested the onStart() and onEnded() hooks with console.log(). They surely are fired but all the onStart() ones execute first followed by the End ones. It seems perhaps all the notes continue to play and do not end after the duration value.

**Reply by drscottlobo** [reply]
I'd suggest moving away from organizing your music data by measure. Measures are just a visual aspect of the music, they don't have any effect on what we hear, so you might want to decouple your visual logic from your playback logic. 

But, perhaps rather than skipping that part of the loop like I did, you could add a tag called isSkipped, so the note doesn't play back but it does have a way of highlighting the correct bar with onStart.

**Reply by naushad-jamil** [reply]
Thank you for your input. Agreed they don't have any relation with what we hear but measures because they are modular and small units help us to work in small chunks for learning. And we were only talking about the challenges of visual aspects here. The playback needs to provide a way so we can track measures. Any DAW or music playing software such as musescore they show note by note, measure by measure visually.

Even when you suggested highlighting the note being played - if the note's duration is borrowing duration from future measures and that note is highlighted conceptually it is incorrect when the tied note in the next measure is being played.

I think I may be able to figure out some logic to get his done. By the way any idea why the onStart() functions all gets executed first then the onEnded() ones?

**Reply by naushad-jamil** [reply, marked as answer]
Hi @danigb and @drscottlobo thanks for your input/ideas with my quest here. I was able to figure out a solution to highlight my measures as they are being played. For now since no other method is known I am using the look ahead notes from current and future measures for tied notes and bring the duration from them to current note but I am scheduling callbacks for each measure as I am scheduling the playback notes by computing how long each measure should take. If there are no lags etc. in playing I think this should work fine. I did some quick testing and it seems to be working fine.
